Add support for illumos cross-compilation

am11 opened the pull request:
Summary of changes:

Essential
* add tryrun configuration for illumos cross-compile.
  * note: `src/installer` didn't had one before.
* detect ld, and based on that choose linker flags.
  * in cross-compile from Ubuntu, toolchain uses GNU ld, otherwise when we are compiling on SmartOS and other illumos-based distros, it is Solaris ld. Full `ld --version` output looks like this:
    > ld: Software Generation Utilities - Solaris Link Editors: 5.11-1.1763 (illumos)
* `link` is an `llvm` [concept](http://llvm.org/docs/CommandGuide/llvm-link.html) which does not have a separate equivalent tool in GCC toolchain; `gcc -flto` provides [similar functionality](https://stackoverflow.com/a/18037712/863980). Perhaps `llvm-link` is unused in our builds and can be removed?
  * reason why it wasn't caught with GCC builds before:
    1. we always found a totally different "wrong" tool: [`link(1)`](https://linux.die.net/man/1/link) via system introspection
    2. in rootfs isolation this unrelated tool is not available, and was failing the build.
* add illumos RID to `System.Data.ODBC` project.
  * this is the only project which requires RID of new system, otherwise fails the cross-build of `src/libraries`.
  * rest of the libraries work will follow in separate PR.

Extra
* add `armv7-alpine` in tryrun scripts.
* prefer `${CMAKE_DL_LIBS}` when possible instead of if/else (as already done under `src/installer`).
* remove trailing slash `.tools` in `.gitignore`, so if they are symlinks they don't get added with `git add :/`

Contributes to: #34944.

am11 commented on src/coreclr/src/pal/tests/palsuite/exception_handling/pal_sxs/test1/CMakeLists.txt:
```diff
@@ -18,18 +18,18 @@ convert_to_absolute_path(DEF_SOURCES1 ${DEF_SOURCES1})
 generate_exports_file(${DEF_SOURCES1} ${EXPORTS_FILE1})
 
-if(CLR_CMAKE_TARGET_LINUX OR CLR_CMAKE_TARGET_FREEBSD OR CLR_CMAKE_TARGET_NETBSD)
+if(CLR_CMAKE_TARGET_LINUX OR CLR_CMAKE_TARGET_FREEBSD OR CLR_CMAKE_TARGET_NETBSD OR LD_GNU)
```
I think we can work out the remaining ld flavors and simplify such conditions based on linker in use; instead of listing each platform individually (as the enclosing code in these block is related to linker options).

am11 commented:
cc @janvorli, @jclulow, with this change, the cross-compiled product succeeds `dotnet helloworld.dll` with coreclr runtime on SmartOS 2020 x64. mono runtime work is done separately: #37560.

am11 commented:
@janvorli, could you please take a look?

janvorli commented on src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt:
```diff
@@ -43,11 +43,11 @@ else(CLR_CMAKE_HOST_WIN32)
         set(END_WHOLE_ARCHIVE -Wl,--no-whole-archive)
 
-        if(CLR_CMAKE_TARGET_SUNOS)
+        if(CLR_CMAKE_TARGET_SUNOS AND LD_SOLARIS)
```
Since you are touching this and we have the same code at many places, would you mind extracting this as a cmake function and using it everywhere?

am11 replied:
Extracted it in `set_exports_linker_option` function.

janvorli commented on src/coreclr/tryrun.cmake:
```diff
@@ -73,6 +72,26 @@ if(TARGET_ARCH_NAME MATCHES "^(armel|arm|arm64|x86)$" OR FREEBSD)
+    set_cache_value(HAVE_COMPATIBLE_ILOGBNAN_EXITCODE 0)
+    set_cache_value(HAVE_COMPATIBLE_LOG10_EXITCODE 1)
+    set_cache_value(HAVE_COMPATIBLE_LOG_EXITCODE 1)
+    set_cache_value(HAVE_LARGE_SNPRINTF_SUPPORT 1)
```
It seems these three should have the `_EXITCODE` at the end of the symbol name too, shouldn't they?

am11 replied:
That's correct. also fixed the FreeBSD case above.

am11 replied:
@janvorli, for some reason these three and the FreeBSD one also work with without `_EXITCODE` (I am not sure how, since all the `runnable` cases must have EXITCODE as the common knowledge dictates).

However, when I changed it to _EXITCODE, I forgot to invert the values, as the outcome is oppose to what the value represent. Caught it in the CI run on a public server.

Will fix it in a follow up PR.

am11 replied:
Fixed as part of #37824.

janvorli commented on eng/native/configurecompiler.cmake:
```diff
@@ -577,4 +577,18 @@ else (CLR_CMAKE_HOST_WIN32)
+    if("${ldVersionOutput}" MATCHES "GNU ld")
+        set(LD_GNU 1)
+    elseif("${ldVersionOutput}" MATCHES "GNU gold")
+        set(LD_GOLD 1)
```
I think we should identify both ld and ld gold as LD_GNU. We compile with ld gold explicitly when CLR_CMAKE_PGO_INSTRUMENT is set.

am11 replied:
Merged gold and GNU branches.

am11 commented on src/coreclr/src/dlls/dbgshim/CMakeLists.txt:
```diff
@@ -26,24 +26,6 @@ else(CLR_CMAKE_TARGET_WIN32)
     generate_exports_file(${DEF_SOURCES} ${EXPORTS_FILE})
 endif(CLR_CMAKE_TARGET_WIN32)
 
-if(CLR_CMAKE_HOST_LINUX OR CLR_CMAKE_HOST_FREEBSD OR CLR_CMAKE_HOST_NETBSD OR CLR_CMAKE_HOST_SUNOS)
```
Ops, I typed it but then it got deleted with the rest of the block. Thanks for the catch.

am11 commented on eng/native/functions.cmake:
```diff
@@ -162,6 +162,28 @@ function(preprocess_compile_asm)
   set(${COMPILE_ASM_OUTPUT_OBJECTS} ${ASSEMBLED_OBJECTS} PARENT_SCOPE)
 endfunction()
 
+function(set_exports_linker_option exports_filename)
```
Moved the extra stuff out. Now it is a single-responsibility function.

am11 committed:
Add support for illumos cross-compilation

am11 commented:
Windows arm failure is https://github.com/dotnet/runtime/issues/38121.

am11 commented:
@janvorli, thanks for the review. I have addressed the feedback. Some blocks were moved under `else(CLR_CMAKE_TARGET_WIN32)` to avoid the platform check again, therefore easier to follow the diff without a whitespace change: https://github.com/dotnet/runtime/pull/37753/files?w=1.
